src/mender-update/cli/actions.cpp

```diff
@@ -161,68 +161,89 @@ static error::Error ResultHandler(standalone::ResultAndError result) {
-		if (ResultContains(result.result, Result::FailedInPostCommit)) {
-			cout << "Installed, but one or more post-commit steps failed." << endl;
-		} else if (
-			ResultContains(result.result, Result::Downloaded | Result::NoRollback)
-			or ResultContains(result.result, Result::Installed | Result::NoRollback)) {
-			cout
-				<< "Installation failed, and Update Module does not support rollback. System may be in an inconsistent state."
-				<< endl;
-		} else if (ResultContains(result.result, Result::NoRollback)) {
-			cout << "Update Module does not support rollback." << endl;
-		} else if (
-			ResultContains(result.result, Result::Downloaded | Result::RollbackFailed)
-			or ResultContains(result.result, Result::Installed | Result::RollbackFailed)) {
-			cout
-				<< "Installation failed, and rollback also failed. System may be in an inconsistent state."
-				<< endl;
-		} else if (
-			ResultContains(result.result, Result::Downloaded | Result::RolledBack)
-			or ResultContains(result.result, Result::Installed | Result::RolledBack)) {
-			cout << "Installation failed. Rolled back modifications." << endl;
-		} else if (ResultContains(result.result, Result::RollbackFailed)) {
-			cout << "Rollback failed. System may be in an inconsistent state." << endl;
-		} else if (ResultContains(result.result, Result::CleanupFailed)) {
-			cout << "Installed, but one or more post-commit steps failed." << endl;
-		} else {
-			stringstream ss;
-			ss << std::hex << static_cast<int>(result.result);
-			log::Error("Unexpected result value: 0x" + ss.str());
+	using r = Result;
+	auto contains = [&result](r val) {
+		return ResultContains(result.result, val);
+	};
+	auto none_of = [&result](r val) {
+		return ResultNoneOf(result.result, val);
+	};
+	auto add = [](string &str, const string &content) {
+		if (str.size() > 0) {
+			str += " ";
 		}
-	} else if (ResultIs(result.result, Result::RolledBack)) {
-		cout << "Rolled back." << endl;
-	} else if (ResultIs(result.result, Result::NoUpdateInProgress)) {
-		cout << "No update in progress." << endl;
-	} else if (ResultIs(result.result, Result::Downloaded)) {
-		cout << "Streamed to storage, but not installed/enabled." << endl;
-	} else if (ResultContains(result.result, Result::Installed | Result::Committed)) {
-		cout << "Installed and committed." << endl;
-	} else if (ResultContains(result.result, Result::Installed)) {
-		cout << "Installed, but not committed." << endl;
-		cout << "Use 'commit' to update, or 'rollback' to roll back the update." << endl;
-	} else if (ResultContains(result.result, Result::Committed)) {
-		cout << "Committed." << endl;
-	} else {
-		stringstream ss;
-		ss << std::hex << static_cast<int>(result.result);
-		log::Error("Unexpected result value: 0x" + ss.str());
+		str += content;
+	};
+
+	string operation_done;
+	string operation_failure;
+
+	// For failure case, include which attempted operation failed.
+	if (contains(r::DownloadFailed)) {
+		operation_failure = "Streaming failed.";
```

Comment:
True, we don't make a distinction, currently. The question is whether we *want* to make a distinction. Back when we discussed what to call the command (before the `--stop-after` flag was proposed, when `stream` would be its own command), we chose the word "stream" to avoid the misconception that the artifact is being stored on the device first, and then installed. Even if it is temporarily stored in the Update Module File Tree, it is all part of the same overall operation. This is different from apt, where you can say `apt download emacs`, and then install it completely independently afterwards using `dkpg -i emacs.deb`. So it was an attempt at making it clearer that streaming and installing are strongly connected operations, and even if it's temporarily stored, it doesn't really change that.

That being said, I don't really have very strong opinion, but I think I'm leaning towards keeping it as is. Does it make sense?

Reply:
You convinced me. Let's keep _stream_